src/Microsoft.OData.Core/Uri/ODataUriConversionUtils.cs

```diff
@@ -512,6 +512,12 @@ internal static object CoerceTemporalType(object primitiveValue, IEdmPrimitiveTy
                         return new DateTimeOffset(dateValue.Year, dateValue.Month, dateValue.Day, 0, 0, 0, new TimeSpan(0));
                     }
 
+                    if (primitiveValue is DateOnly dateOnly)
+                    {
+                        var dateValue = (Date)dateOnly;

```

Comment:
Why is this coercion necessary?

Reply:
I noticed we have the logic for 'Date', so I added a similar logic for DateOnly. #Resolved